swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,927 @@
+            state: self.phase.state,
+        }
+    }
+
+    pub fn without_outbound_handler(self) -> FromFnProto<TInbound, Void, Void, TState> {
+        FromFnProto {
+            protocol: self.protocol,
+            on_new_inbound: self.phase.inbound_handler,
+            on_new_outbound: Box::new(|_, _, _, _, info| void::unreachable(info)),
+            inbound_streams_limit: self.phase.max_inbound,
+            pending_outbound_streams_limit: 0,
+            state: self.phase.state,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum OutEvent<I, O, OpenInfo> {
+    InboundEmitted(I),
+    OutboundEmitted(O),
+    FailedToOpen(OpenError<OpenInfo>),
+}
+
+#[derive(Debug)]
+pub enum OpenError<OpenInfo> {
+    /// The time limit for the negotiation handshake was exceeded.
+    Timeout(OpenInfo),
+    /// We have hit the configured limit for the maximum number of pending substreams.
+    LimitExceeded {
+        open_info: OpenInfo,
+        pending_substreams: usize,
+    },
+    /// The remote does not support this protocol.
+    Unsupported {
+        open_info: OpenInfo,
+        protocol: &'static str,
+    },
+    NegotiationFailed(OpenInfo, ProtocolError),
+}
+
+/// A wrapper for state that is shared across all connections.
+///
+/// Any update to the state will "automatically" be relayed to all connections, assuming this struct
+/// is correctly wired into your [`NetworkBehaviour`](crate::swarm::NetworkBehaviour).
+///
+/// This struct implements an observer pattern. All registered connections will receive updates that
+/// are made to the state.
+pub struct Shared<T> {
+    inner: T,
+
+    shared: Arc<T>,
+
+    dirty: bool,
+    waker: Option<Waker>,
+    connections: HashSet<(PeerId, ConnectionId)>,
+    pending_update_events: VecDeque<(PeerId, ConnectionId, Arc<T>)>,
+}
+
+impl<T> Shared<T>
+where
+    T: Clone,
+{
+    pub fn new(state: T) -> Self {
+        Self {
+            inner: state.clone(),
+            shared: Arc::new(state),
+            dirty: false,
+            waker: None,
+            connections: HashSet::default(),
+            pending_update_events: VecDeque::default(),
+        }
+    }
+
+    pub fn register_connection(&mut self, peer_id: PeerId, id: ConnectionId) {
+        self.connections.insert((peer_id, id));
+    }
+
+    pub fn unregister_connection(&mut self, peer_id: PeerId, id: ConnectionId) {
+        self.connections.remove(&(peer_id, id));
+    }
+
+    pub fn poll<TOut, THandler, TOpenInfo>(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<NetworkBehaviourAction<TOut, THandler, InEvent<T, TOpenInfo>>>
+    where
+        THandler: IntoConnectionHandler,
+    {
+        if self.dirty {
+            self.shared = Arc::new(self.inner.clone());
+
+            self.pending_update_events = self
+                .connections
+                .iter()
+                .map(|(peer_id, conn_id)| (*peer_id, *conn_id, self.shared.clone()))
+                .collect();
```

Comment:
This erases previously undelivered wake-ups — which is fine unless it happens in a tight loop.

Reply:
The old events are useless though because they contain stale state. I am not sure if there is a better way?

Reply:
Right, old events are useless. Perhaps I am overthinking this, my worry is that there may be a situation where only the first (few) connections are ever notified because updates keep coming in — a fairness concern. This would be solved by randomising the notification order, or by always finishing one batch of notifications (with values updated during the process!) before starting the next. The second is deterministic but more work to implement.

Getting these things wrong usually has low chance of severe impact, but when the impact happens it is really difficult for downstream code to find the bug.

Reply:
I think that we currently drain events from the `NetworkBehaviour` before we inject new ones into it so I don't think that this would happen but I'd have to double check that.